In [43]:
import numpy as np
import pandas as pd

In [44]:
# import the dataset
df = pd.read_csv("sample_0_ptrac.csv");
df

Type          X          Y          Z         U         V         W  \
0    1000   0.000000   0.000000   0.000000  0.508477  0.473278  0.719347   
1    3000  26.153789  24.343315  37.000000  0.508477  0.473278  0.719347   
2    4000  28.567736  26.590158  40.415032  0.508477  0.473278  0.719347   
3    5000  28.567736  26.590158  40.415032  0.508477  0.473278  0.719347   
4    2000  28.567736  26.590158  40.415032  0.642856 -0.725381  0.246084   
..    ...        ...        ...        ...       ...       ...       ...   
427  5000  42.260018  11.133571  45.659893  0.956422  0.291984  0.001599   
428  2000  42.259198  11.136909  45.658301  0.320801  0.453103  0.831736   
429  5000  42.259198  11.136909  45.658302 -0.043768 -0.699589  0.713204   
430  2000  42.258623  11.139260  45.657183  0.096900  0.258149  0.961233   
431  5000  42.258623  11.139261  45.657186 -0.333001  0.020514  0.942703   

        Energy  Weight  Cell  Material      Time  
0    14.000000     1.0  30.0       6.0  0.000000  
1    14.000000     1.0  20.0       1.0  1.004945  
2    14.000000     1.0  20.0       1.0  1.097699  
3    14.000000     1.0  20.0       1.0  1.097699  
4     2.429137     1.0  20.0       1.0  1.097699  
..         ...     ...   ...       ...       ...  
427   0.000982     1.0  20.0       1.0  1.168771  
428   0.001087     1.0  20.0       1.0  1.168757  
429   0.000993     1.0  20.0       1.0  1.168757  
430   0.001121     1.0  20.0       1.0  1.168748  
431   0.000729     1.0  20.0       1.0  1.168748  

[432 rows x 12 columns]

In [45]:
df.columns

Index(['Type', 'X', 'Y', 'Z', 'U', 'V', 'W', 'Energy', 'Weight', 'Cell',
       'Material', 'Time'],
      dtype='object')

In [46]:
df['Type'].unique()

array([1000, 3000, 4000, 5000, 2000])

In [47]:
df['Type'].value_counts()

Type
5000    204
2000    203
4000     23
1000      1
3000      1
Name: count, dtype: int64

Starts with one particle.  
203 new particles were created.   
204 (all particles) were destroyed.  

In [48]:
is_increasing = all(df['Time'].iloc[i] > df['Time'].iloc[i-1] for i in range(1, len(df)))

if is_increasing:
    print("The time column is strictly increasing.")
else:
    print("The time column is not strictly increasing.")

The time column is not strictly increasing.


In [49]:
# Check if the time column is strictly decreasing
is_decreasing = all(df['Time'].iloc[i] < df['Time'].iloc[i-1] for i in range(1, len(df)))

if is_decreasing:
    print("The time column is strictly decreasing.")
else:
    print("The time column is not strictly decreasing.")

The time column is not strictly decreasing.


In [50]:
# Initialize variables
datasets = []
current_dataset = []

# Iterate over the DataFrame
for index, row in df.iterrows():
    event_type = row['Type']
    
    # Check if the current row is the start of a new dataset
    if event_type == 2000:
        if current_dataset:  # Append the current dataset if it's not empty
            datasets.append(pd.DataFrame(current_dataset))
            current_dataset = []  # Reset the current dataset
        
    # Append the current row to the current dataset
    current_dataset.append([row['Type'], row['X'], row['Y'], row['Z'], row["Time"]])
    
# Append the last dataset
if current_dataset:
    datasets.append(pd.DataFrame(current_dataset))

# Print the number of datasets created
print("Number of datasets:", len(datasets))


Number of datasets: 204


In [51]:
# Print each dataset
for idx, dataset in enumerate(datasets, start=1):
    print(f"Dataset {idx}:")
    print(dataset)
    print("\n") 

Dataset 1:
        0          1          2          3         4
0  1000.0   0.000000   0.000000   0.000000  0.000000
1  3000.0  26.153789  24.343315  37.000000  1.004945
2  4000.0  28.567736  26.590158  40.415032  1.097699
3  5000.0  28.567736  26.590158  40.415032  1.097699


Dataset 2:
         0          1          2          3         4
0   2000.0  28.567736  26.590158  40.415032  1.097699
1   4000.0  42.257788  11.142679  45.655559  1.168734
2   4000.0  44.316991  10.528963  45.542828  1.175911
3   4000.0  45.787397  11.055095  47.935630  1.185442
4   4000.0  38.756944  17.174070  48.410751  1.216572
5   4000.0  36.259014  19.504108  49.087581  1.228188
6   4000.0  37.246132  17.249322  51.136169  1.238870
7   4000.0  42.338193  19.694256  45.077795  1.266500
8   4000.0  42.682331   8.300546  38.348026  1.310654
9   4000.0  45.101685  10.208199  38.599819  1.320966
10  4000.0  44.436272   8.738580  38.901607  1.326440
11  4000.0  43.635166   8.642429  38.132947  1.330157
12  4000.

In [52]:
# printing the datasets where the particle dies as soon as it is created
for idx, dataset in enumerate(datasets, start=1):
    if len(dataset) == 2:
        print(f"Dataset {idx}:")
        print(dataset)
        print("\n")
        
        # Calculate coordinate differences and time difference
        x_diff = dataset.iloc[1][1] - dataset.iloc[0][1]
        y_diff = dataset.iloc[1][2] - dataset.iloc[0][2]
        z_diff = dataset.iloc[1][3] - dataset.iloc[0][3]
        time_diff = dataset.iloc[1][4] - dataset.iloc[0][4]
        
        # Print coordinate differences and time difference
        print("Coordinate Differences:")
        print(f"  X-coordinate Difference: {x_diff:.6f}")
        print(f"  Y-coordinate Difference: {y_diff:.6f}")
        print(f"  Z-coordinate Difference: {z_diff:.6f}")
        print(f"  Time Difference: {time_diff:.6f}")
        print("\n")

Dataset 3:
        0          1         2          3         4
0  2000.0  43.982648  8.806252  37.874848  1.336365
1  5000.0  43.981783  8.805031  37.875197  1.336406


Coordinate Differences:
  X-coordinate Difference: -0.000864
  Y-coordinate Difference: -0.001222
  Z-coordinate Difference: 0.000349
  Time Difference: 0.000042


Dataset 4:
        0          1         2          3         4
0  2000.0  43.981398  8.805443  37.875640  1.336385
1  5000.0  43.981407  8.805406  37.875622  1.336387


Coordinate Differences:
  X-coordinate Difference: 0.000008
  Y-coordinate Difference: -0.000036
  Z-coordinate Difference: -0.000017
  Time Difference: 0.000002


Dataset 5:
        0          1         2          3         4
0  2000.0  43.998368  8.794892  37.929746  1.336171
1  5000.0  43.998319  8.794924  37.929838  1.336175


Coordinate Differences:
  X-coordinate Difference: -0.000048
  Y-coordinate Difference: 0.000032
  Z-coordinate Difference: 0.000092
  Time Difference: 0.000004


Da

In [53]:
# printing the datasets that have more than 2 rows
for idx, dataset in enumerate(datasets, start=1):
    if len(dataset) > 2:
        print(f"Dataset {idx}:")
        print(dataset)
        print("\n")

Dataset 1:
        0          1          2          3         4
0  1000.0   0.000000   0.000000   0.000000  0.000000
1  3000.0  26.153789  24.343315  37.000000  1.004945
2  4000.0  28.567736  26.590158  40.415032  1.097699
3  5000.0  28.567736  26.590158  40.415032  1.097699


Dataset 2:
         0          1          2          3         4
0   2000.0  28.567736  26.590158  40.415032  1.097699
1   4000.0  42.257788  11.142679  45.655559  1.168734
2   4000.0  44.316991  10.528963  45.542828  1.175911
3   4000.0  45.787397  11.055095  47.935630  1.185442
4   4000.0  38.756944  17.174070  48.410751  1.216572
5   4000.0  36.259014  19.504108  49.087581  1.228188
6   4000.0  37.246132  17.249322  51.136169  1.238870
7   4000.0  42.338193  19.694256  45.077795  1.266500
8   4000.0  42.682331   8.300546  38.348026  1.310654
9   4000.0  45.101685  10.208199  38.599819  1.320966
10  4000.0  44.436272   8.738580  38.901607  1.326440
11  4000.0  43.635166   8.642429  38.132947  1.330157
12  4000.

So most of the particles are terminates as soon as they are created almost at the same coordinates.  
We have five cases where the particle traverses before it is terminated.